### PR TITLE
[LIVY-586] Fix batch state from starting to dead when startup fail

### PR DESCRIPTION
**Pull request by runzhiwang**
## What changes were proposed in this pull request?

Fix batch state from starting to dead when startup fail

## How was this patch tested?

Create batch session with error parameter, the batch state will turn from starting to dead.

### CODE REVIEW

File: server/src/main/scala/org/apache/livy/utils/SparkYarnApp.scala [hunk 1]
```diff
@@ -169,6 +169,10 @@ class SparkYarnApp private[utils] (
       appTag: String,
       pollInterval: Duration,
       deadline: Deadline): ApplicationId = {
+    if (process.isDefined && !process.get.isAlive && process.get.exitValue() != 0) {
```
Comment (yiheng) [inline server/src/main/scala/org/apache/livy/utils/SparkYarnApp.scala]: create a unit test to cover this case?

Comment (runzhiwang) [reply server/src/main/scala/org/apache/livy/utils/SparkYarnApp.scala]: okay

Comment (runzhiwang) [reply server/src/main/scala/org/apache/livy/utils/SparkYarnApp.scala]: @yiheng UT has been added.

**Comment by yiheng** [comment]
+1

File: server/src/main/scala/org/apache/livy/utils/SparkYarnApp.scala (hunk 1 above)
Comment (jerryshao) [reply server/src/main/scala/org/apache/livy/utils/SparkYarnApp.scala]: Seems a bit weird to add check here in this method, maybe we can add such logic in `yarnAppMonitorThread`.

Comment (runzhiwang) [reply server/src/main/scala/org/apache/livy/utils/SparkYarnApp.scala]: @jerryshao The process is alive when the yarnAppMonitorThread was created, so I cannot check it in the begining of yarnAppMonitorThread. The process turns to dead when getAppIdFromTag which always fail and cost 120 seconds, so I check it in getAppIdFromTag  to make the state change as soon as possible.

Comment (jerryshao) [reply server/src/main/scala/org/apache/livy/utils/SparkYarnApp.scala]: I see. Can we please make such check as a method, I saw there's another place you changed previously which has the same check.

File: server/src/main/scala/org/apache/livy/utils/SparkYarnApp.scala [hunk 2]
```diff
@@ -151,6 +151,13 @@ class SparkYarnApp private[utils] (
     }
   }
 
+  private def checkProcessErrExit(): Boolean = {
+    if (process.isDefined && !process.get.isAlive && process.get.exitValue() != 0) {
+      return true
+    }
+    return false
+  }
```
Comment (jerryshao) [inline server/src/main/scala/org/apache/livy/utils/SparkYarnApp.scala]:
```
private def isProcessErrExit(): Boolean = {
    process.isDefined && !process.get.isAlive && process.get.exitValue() != 0
}
```

Comment (runzhiwang) [reply server/src/main/scala/org/apache/livy/utils/SparkYarnApp.scala]: okay

Comment (runzhiwang) [reply server/src/main/scala/org/apache/livy/utils/SparkYarnApp.scala]: @jerryshao I have updated the commit.

**Pull request closed without merging**